library/std/src/collections/hash/map.rs

```diff
@@ -279,7 +279,18 @@ impl<K, V, S> HashMap<K, V, S> {
     /// ```
     #[inline]
     #[stable(feature = "hashmap_build_hasher", since = "1.7.0")]
-    #[rustc_const_unstable(feature = "const_collections_with_hasher", issue = "102575")]
+    #[cfg_attr(
+        bootstrap,
+        rustc_const_unstable(feature = "const_collections_with_hasher", issue = "102575")
+    )]
+    #[cfg_attr(
+        not(bootstrap),
+        rustc_const_stable(
+            feature = "const_collections_with_hasher",
+            since = "CURRENT_RUSTC_VERSION"
+        )
+    )]
+    #[rustc_allow_const_fn_unstable(any)]
```

Comment:
So this would allow this function to call *any* unstable function? That's not great, we specifically want to keep an eye on what is (recursively) exposed to `const` on stable, and with such an "any" handling that's not really possible.

Reply:
In fact this calls into hashbrown, doesn't it? We'd want to ensure that hashbrown only uses const-stable features then. const nightly features are much more prone to making things fail-to-build when they get changed, which is why we have a more strict stability check than for regular non-const functions.

I'm not sure what is the best way to do that. This might have to wait until Oli is back from vacation, it's a pretty fundamental change to our const stabilization policy.

Reply:
> So this would allow this function to call any unstable function?

No, only unmarked unstable functions. I should add a test for this (and change the name as well)

> We'd want to ensure that hashbrown only uses const-stable features then. 

This already happens because hashbrown's functions are `const` on stable (so only have access to things that have been const-stabilized).

However, because deps-of-std are build with `-Zforce-unstable-if-unmarked`, the compiler believes them to be unstably-const.

> Ideally we'd be able to, and be required to, mark functions as rustc_const_stable in hashbrown itself; 

This could be accomplished with a load of

```
#[cfg(feature = "rustc_dep_of_std", rustc_const_stable(...)]
```

but it would require a load of ugly changes to hashbrown, as this would also require them to mark all public functions as stable (because currently stability and const-stability are under the same `#![feature(staged_api)]`. Spliting that up would be a lot of work for very little benefit.

Reply:
> This already happens because hashbrown's functions are const on stable (so only have access to things that have been const-stabilized).

I see a lot of `feature` flags in hashbrown.  How can we be sure that the hashbrown built for the standard library still only uses stable features?

Reply:
> How can we be sure that the hashbrown built for the standard library still only uses stable const features?

I supose hashbrown could change to `#[cfg_attr(feature="rustc_dep_of_std"), feature(const_super_unstable_and_broken)]`, and we wouldn't mechanicly know. I think the best way would be to ask them not to do that. Given that hashbrown needs to be able to provide a const-stable implementation of new, I'm not concerned about being backed into a corner where there's no way to implement new with stable const features.

Reply:
Oh sure I am not suggesting we have to assume hashbrown to try and subvert this. :)

But I'd prefer if this was automatically tested. `rustc-dep-of-std` enables `nightly` which enables some critical features that could accidentally expose things we don't want to expose yet. When reviewing a hashbrown PR it seems easy to miss that a new intrinsic being added is (a) const-unstable and (b) now exposed to const.

Reply:
`core_intrinsics` seems to only be used for
```rust
#[cfg(feature = "nightly")]
use core::intrinsics::{likely, unlikely};
```
Can we move those under a different feature so that hashbrown can use them but not accidentally use anything else?

Reply:
> What are "unmarked" unstable functions?

See https://github.com/rust-lang/rust/blob/b10cfcd65fd7f7b1ab9beb34798b2108de003452/compiler/rustc_const_eval/src/transform/check_consts/check.rs#L984-L989

When a function is unstable, but not either const-stable or const-unstable. This is true for hashbrown, despite it not having any stability attributes, as it is built with `-Zforce-unstable-if-unmarked` such that it isn't availibe to normal crates, despite being available in the sysroot

https://github.com/rust-lang/rust/blob/b10cfcd65fd7f7b1ab9beb34798b2108de003452/compiler/rustc_passes/src/stability.rs#L635-L651

Reply:
> Can we move those under a different feature so that hashbrown can use them but not accidentally use anything else?

Don't see why not. Although it looks like the pattern with intrinsic is to add wrappers that don't live in intrinsic, and put them on a separate feature and maybe stabilize them.

> When reviewing a hashbrown PR it seems easy to miss that a new intrinsic being added is (a) const-unstable and (b) now exposed to const.

`#![feature(core_intrinsics)]` doesn't allow using them in const-contexts, that's on a seperate feature. While it's not impossible, it'd be much more likely (ha) to be flagged in review.

<https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=43a21473dbd0af03612e0805e2a847f2>

Reply:
> When a function is unstable, but not either const-stable or const-unstable. This is true for hashbrown, despite it not having any stability attributes, as it is built with -Zforce-unstable-if-unmarked such that it isn't availibe to normal crates, despite being available in the sysroot

Okay I see. I also see this comment
```
                // FIXME(ecstaticmorse); For compatibility, we consider `unstable` callees that
                // have no `rustc_const_stable` attributes to be const-unstable as well. This
                // should be fixed later.
```
@ecstatic-morse I wonder what you meant by "fixing" this, what would be the more correct behavior?

IMO these should be treated like those with explicit `rustc_const_unstable` except they have no feature gate.

> Don't see why not. Although it looks like the pattern with intrinsic is to add wrappers that don't live in intrinsic, and put them on a separate feature and maybe stabilize them.

Yes. But those will often not be const at all, or they are const unstable and hashbrown shouldn't use that feature ideally.

Hm... we could raise an error when an "unmarked" const function calls a const-unstable function? That should guarantee that when we allow `rustc_allow_const_fn_unstable(unmarked)` this can never call a function explicitly marked as const-unstable, not even transitively?

Reply:
> we could raise an error when an "unmarked" const function calls a const-unstable function?

We already do this! I've added a test for this.

Also, I've renamed the attribute to `#[rustc_allow_const_fn_unstable(any_unmarked)]`, so it's clearer that you can't call arbitrary unstableley-const fns.